connector/src/main/scala/com/microsoft/kusto/spark/authentication/DeviceAuthentication.scala

```diff
@@ -1,115 +1,63 @@
-    val deviceCode = if (userDeviceCode.isDefined) userDeviceCode.get else acquireDeviceCode(clusterUrl, authority)
-    val aadAuthorityUri = s"https://login.microsoftonline.com/$authority"
-    val service: ExecutorService =
-      Executors.newSingleThreadExecutor
-    val context: AuthenticationContext =  new AuthenticationContext(aadAuthorityUri, true, service)
-
-    var text: String = null
-    Try {
-      val clipboard = Toolkit.getDefaultToolkit.getSystemClipboard
-      val dataFlavor = DataFlavor.stringFlavor
-      if (clipboard.isDataFlavorAvailable(dataFlavor)) {
-        text = clipboard.getData(dataFlavor).asInstanceOf[String]
-      }
-      clipboard.setContents(new StringSelection(deviceCode.getUserCode), null)
+  def acquireNewAccessTokenAsync(): CompletableFuture[IAuthenticationResult] = {
+    val deviceCodeConsumer: Consumer[DeviceCode] = (deviceCode: DeviceCode) => {
+      this.deviceCode = Some(deviceCode)
+      expiresAt = Some(System.currentTimeMillis + deviceCode.expiresIn() * 1000)
+      System.out.println(deviceCode.message())
     }
 
-    println(deviceCode.getMessage +
-      (if (!Desktop.isDesktopSupported) deviceCode
-      else " device code is already copied to clipboard - just press ctrl+v in the web"))
-    if (Desktop.isDesktopSupported) Desktop.getDesktop.browse(new URI(deviceCode.getVerificationUrl))
-    val result = waitAndAcquireTokenByDeviceCode(deviceCode, context)
-    if(text != null) {
-      Try {
-        val clipboard = Toolkit.getDefaultToolkit.getSystemClipboard
-        clipboard.setContents(new StringSelection(text), null)
-      }
-    }
-    if (result == null) throw new ServiceUnavailableException("authentication result was null")
-    result.getAccessToken
+    val deviceCodeFlowParams: DeviceCodeFlowParameters = DeviceCodeFlowParameters.builder(scopes, deviceCodeConsumer).build
+    clientApplication.acquireToken(deviceCodeFlowParams)
   }
 
-  @throws[InterruptedException]
-  private def waitAndAcquireTokenByDeviceCode(deviceCode: DeviceCode, context: AuthenticationContext): AuthenticationResult = {
-    var timeout = 15 * 1000
-    var result: AuthenticationResult = null
-
-    // Logging is set to Fatal as root
-    val prevLevel = Logger.getLogger(classOf[AuthenticationContext]).getLevel
-    Thread.sleep(5000)
-    Logger.getLogger(classOf[AuthenticationContext]).setLevel(Level.FATAL)
-    while ( result == null && timeout > 0) {
-      try {
-        result = context.acquireTokenByDeviceCode(deviceCode, null).get
-
-      } catch {
-        case e: ExecutionException =>
-          Thread.sleep(1000)
-          timeout -= 1000
+  def refreshIfNeeded(): Unit = {
+    if (deviceCode.isEmpty || expiresAt.get <= System.currentTimeMillis) {
+      val oldDeviceCode = this.deviceCode
+      awaitAuthentication = Some(acquireNewAccessTokenAsync())
+      var awaitTime = NEW_DEVICE_CODE_FETCH_TIMEOUT
```

Comment:
New to scala here, but shouldn't this be a val?

Reply:
no
as it changes later